engine/graphics/font.cpp

```diff
@@ -0,0 +1,171 @@
+#include "font.h"
+
+#include "image.h"
+#include "logger.h"
+#include "utils.h"
+
+#define STB_TRUETYPE_IMPLEMENTATION
+#include <stb_truetype.h>
+
+#include <fstream>
+
+namespace pong
+{
+
+// The font will load these characters by default.
+// If a character is not in this list, it will be loaded the first time it is used.
+static const std::string DEFAULT_FONT_CHARACTERS =
+    "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*()-=_+[]{}\\|;:'\",<.>/?`~ ";
+
+static constexpr int PIXEL_LINE_HEIGHT = 256;
+
+bool Font::LoadFont(const std::string& fontName, const std::string& fontPath)
+{
+    std::ifstream file(fontPath, std::ios::binary);
+    if (!file)
+    {
+        LogError("Failed to open true type font file: {} - {}", fontPath, fontName);
+        ASSERT(false);
+        return false;
+    }
+
+    mFontData = std::vector<unsigned char>(std::istreambuf_iterator<char>(file), std::istreambuf_iterator<char>());
+    ASSERT(stbtt_InitFont(&mFontInfo, mFontData.data(), 0) != 0);
+
+    mPixelScale = stbtt_ScaleForPixelHeight(&mFontInfo, static_cast<float>(PIXEL_LINE_HEIGHT));
+
+    int unscaledAscent, unscaledDescent, unscaledLineGap;
+    stbtt_GetFontVMetrics(&mFontInfo, &unscaledAscent, &unscaledDescent, &unscaledLineGap);
+
+    mAscent  = unscaledAscent  * mPixelScale;
+    mDescent = unscaledDescent * mPixelScale;
+    mLineGap = unscaledLineGap * mPixelScale;
+
+    for (const char& character : DEFAULT_FONT_CHARACTERS)
+    {
+        LoadCharacter(character);
+    }
+
+    return true;
+}
+
+FontCharacter Font::LoadCharacter(const char character)
+{
+    FontCharacter fontCharacter {};
+
+    int unscaledGlpyhWidth = 0;
+    int unscaledLeftSideBearing = 0;
+    stbtt_GetCodepointHMetrics(&mFontInfo, character, &unscaledGlpyhWidth, &unscaledLeftSideBearing);
+
+    fontCharacter.mGlyphWidth = unscaledGlpyhWidth * mPixelScale;
+    fontCharacter.mLeftSideBearing = unscaledLeftSideBearing * mPixelScale;
+
+    int xCoord1 = 0;
+    int yCoord1 = 0;
+    int xCoord2 = 0;
+    int yCoord2 = 0;
+    stbtt_GetCodepointBitmapBox(&mFontInfo, character, mPixelScale, mPixelScale,
+                                &xCoord1, &yCoord1, &xCoord2, &yCoord2);
+
+    const int charWidth  = xCoord2 - xCoord1;
+    const int charHeight = yCoord2 - yCoord1;
+
+    std::vector<unsigned char> alphaTexture { 0 };
```

Comment:
Something to be very mindful of is that std::vector has a bunch of dumb initializers, two of which are very easy to mess up:

```
std::vector<int> b{3}; // creates a 1-element vector holding {3}
std::vector<int> d(3); // creates a 3-element vector holding {0, 0, 0}
```

If your intention was to create a size 1 vector with the only element being initialized to zero, you succeeded.

(just pointing this out as a frequent bug that bites every now and again, it's constantly called out in C++ books and was probably a design mistake)